platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -347,11 +346,17 @@ private void doVisitDependencies(ExcludeSpec resolutionFilter, StrictVersionCons
     private void visitNewAndInvalidatedDependencies(ExcludeSpec resolutionFilter, StrictVersionConstraints ancestorsStrictVersions, Collection<EdgeState> discoveredEdges) {
         // Visit any constraints that were previously pending, but are no longer pending.
         if (upcomingNoLongerPendingConstraints != null && potentiallyActivatedConstraints != null) {
-            for (ModuleIdentifier module : upcomingNoLongerPendingConstraints) {
-                Collection<DependencyState> dependencyStates = potentiallyActivatedConstraints.get(module);
+            for (ModuleIdentifier moduleId : upcomingNoLongerPendingConstraints) {
+                Collection<DependencyState> dependencyStates = potentiallyActivatedConstraints.get(moduleId);
                 if (!dependencyStates.isEmpty()) {
-                    for (DependencyState dependencyState : dependencyStates) {
-                        createAndLinkEdgeState(dependencyState, discoveredEdges, resolutionFilter, ancestorsStrictVersions, false);
+                    ModuleResolveState module = resolveState.getModule(moduleId);
+                    if (module.isPending()) {
```

Comment:
See `dependency constraint can be not pending, then pending, then not pending and still participate in resolution`. 

We would notify this node that one of its outgoing constraints needs to be activated, but in between that time and when this node comes up in the queue, the target module of the potential constraint goes back to pending. We previously didn't detect this and then went all the way to trying to attach the edge to the target node in the pending module before realizing this. 

We detect this much earlier and no longer need to try to detach this stale edge during `attachToTargetNodes`